tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala

```diff
@@ -0,0 +1,92 @@
+package at.forsyte.apalache.tla
+
+import at.forsyte.apalache.tla.lir.oper.TlaOper
+import at.forsyte.apalache.tla.lir.{NameEx, OperT1, TlaEx, TlaType1, ValEx}
+
+import scala.language.implicitConversions
+import scalaz._
+
+// Contains classes, typedefs and implicit conversion methods used in the package
+package object typecmp {
+
+  // Builder-thrown exceptions
+  /** Thrown if a TypeComputation finds fault with types */
+  class BuilderTypeException(message: String) extends Exception(message)
+
+  /** Thrown if scope validation finds a clash between multiple instances of the same variable name */
+  class BuilderScopeException(message: String) extends Exception(message)
+
+  /** Type computations return types or TypeExceptions */
+  type TypeComputationReturn = Either[BuilderTypeException, TlaType1]
+
+  /** Builder methods return TlaEx (they throw exceptions) */
+  type BuilderReturn = TlaEx
+
+  /**
+   * A type computation returns the result type of an operator application, given the concrete arguments. In general,
+   * argument types are not enough, because we need concrete fields/indices for records/sequences.
+   */
+  type TypeComputation = Seq[TlaEx] => TypeComputationReturn
+
+  /**
+   * For some (most) operators, we don't need the exact values (just types) of the arguments to determine the result
+   * type. Pure computations are TypeComputations that are associated with such operators.
+   */
+  type PureTypeComputation = Seq[TlaType1] => TypeComputationReturn
+
+  /**
+   * MetaInfo holds all of the information about the internal state of the builder. For now, this just includes the name
+   * scope, but it can be extended in the future, to have the builder perform additional static analysis, e.g.
+   * assignment analysis.
+   *
+   * `nameScope` tracks the types of the variables currently considered as free
+   */
+  final case class MetaInfo(nameScope: Map[String, TlaType1])
+
+  /** An IntenalState is a computation (possibly) mutating some MetaInfo */
+  type InternalState[T] = State[MetaInfo, T]
+
+  /** Builders operate over BuilderWrappers, which hold MetaInfo as state */
+  type BuilderWrapper = InternalState[BuilderReturn]
```

Comment:
```suggestion
  type TBuilderPiece = InternalState[BuilderReturn]
```

Do the users of your API care about the fact that this is a wrapper? It's important that this is a piece of work to be done by the builder. So I would called it a "piece", a "part", etc.

Reply:
`-Instruction`? Since an `InternalState[TlaEx]` defines how to build `TlaEx` on demand?